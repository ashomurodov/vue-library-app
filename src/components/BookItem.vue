<template>
	<div class="book-container">
		<span class="img">
			<router-link :to="'/book/' + props.book.id">
				<img :src="props.book.image" alt="" />
			</router-link>
		</span>
		<div class="bookDetails">
			<h3>
				<router-link style="text-decoration: none; color: black" :to="'/book/' + props.book.id">{{
					props.book.title
				}}</router-link>
			</h3>
			<div class="authors">
				<p v-for="author of props.book.authors" :key="author + Date.now()">
					{{ author }}
				</p>
			</div>
		</div>
	</div>
</template>

<script setup>
import { defineProps } from "vue";

const props = defineProps({
	book: Object,
});
</script>

<style scoped>
* {
	margin: 0;
	padding: 0;
	box-sizing: border-box;
}
.book-container {
	display: flex;
	justify-content: space-between;
	flex-direction: column;
	align-items: center;
	border: 2px solid #2c3e50;
	width: 300px;
	height: 500px;
	background-color: #f3f3f3;
	padding: 20px;
	font-family: sans-serif;
}

@media screen and (max-width: 1790px) {
	.book-container {
		width: 350px;
		height: 470px;
	}
}

@media screen and (max-width: 1680px) {
	.book-container {
		width: 430px;
		height: 570px;
	}

	h3,
	.authors {
		width: 350px;
	}
}

@media screen and (max-width: 1500px) {
	
}

.bookDetails {
	display: flex;
	flex-direction: column;
	gap: 3px;
	align-items: flex-start;
}

.img {
	height: 80%;
	width: 80%;
	background-color: #17bb7f;
}

.img img {
	width: 100%;
	height: 100%;
	object-fit: cover;
	display: flex;
	justify-content: center;
	align-items: center;
}

h3 {
	font-size: 17px;
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
	text-align: start;
	width: 200px; /* Adjust the width to your desired value */
}

.authors {
	display: flex;
	gap: 10px;
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
	width: 200px; /* Adjust the width to your desired value */
}

p {
	font-size: 18px;
	color: rgb(75, 75, 75);
}
</style>
