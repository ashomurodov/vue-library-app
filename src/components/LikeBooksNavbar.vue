<template>
  <div
    class="flex justify-between items-center lg:px-20 px-10 py-5 bg-emerald-100 z-50 sticky w-full top-0"
  >
    <div class="logo">
      <router-link to="/">
        <p class="text-black hidden sm:block lg:text-5xl md:text-3xl text-2xl">
          Back to <span class="text-emerald-600">Menu!</span>
        </p>
        <p class="text-black block sm:hidden text-2xl">
          Ba<span class="text-emerald-600">ck!</span>
        </p>
      </router-link>
    </div>

    <div class="flex justify-center items-center py-3">
      <img
        class="ml-3 cursor-pointer"
        id="logoutimg"
        @click="logOut"
        :src="logOutIcon"
        alt="logout"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { logOut } from "@/utils";
import { SearchBar } from ".";
import logOutIcon from "@/assets/logouticon.png";
</script>
