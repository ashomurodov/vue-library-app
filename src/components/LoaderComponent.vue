<template>
	<div v-if="loading" class="loader">
		<div class="spinner"></div>
	</div>
</template>

<script>
export default {
	props: {
		loading: Boolean,
	},
};
</script>

<style scoped>
.loader {
	display: flex;
	justify-content: center;
	align-items: center;
	height: 100%;
}

.spinner {
	border: 4px solid rgba(10, 40, 31, 0.3);
	border-top: 4px solid #17bb7f;
	border-radius: 50%;
	width: 40px;
	height: 40px;
	animation: spin 1s linear infinite;
}

@keyframes spin {
	0% {
		transform: rotate(0deg);
	}
	100% {
		transform: rotate(360deg);
	}
}
</style>
