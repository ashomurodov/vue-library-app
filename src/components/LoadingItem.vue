<template>
  <div class="grid place-items-center h-full">
    <img :src="loader" alt="loader" />
  </div>
</template>

<script setup lang="ts">
import loader from "@/assets/loader.gif";
</script>
