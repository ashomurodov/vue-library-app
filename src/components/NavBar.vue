<template>
  <div
    class="flex justify-between items-center lg:px-20 px-10 py-5 bg-emerald-100 sticky w-full top-0"
  >
    <div class="logo">
      <router-link to="/">
        <p class="text-black hidden sm:block lg:text-5xl md:text-3xl text-2xl">
          Read<span class="text-emerald-600">Zen!</span>
        </p>
        <p class="text-black block sm:hidden text-2xl">R<span class="text-emerald-600">Z!</span></p>
      </router-link>
    </div>
    <div class="flex justify-center items-center py-3">
      <SearchBar />
      <img class="ml-3 cursor-pointer" @click="handleSubmit" :src="logout" alt="logout" />
    </div>
  </div>
</template>

<script setup lang="ts">
import { SearchBar } from ".";
import logout from "@/assets/logouticon.png";
import router from "@/router";

const handleSubmit = () => {
  const isConfirmed = confirm("Tizimdan chiqishga rozimisiz?");
  if (isConfirmed) {
    localStorage.removeItem("user_data");
    router.push("/login");
  }
};
</script>
