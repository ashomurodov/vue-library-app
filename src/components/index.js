export { default as BookItem } from "./BookItem.vue";
export { default as LoaderComponent } from "./LoaderComponent.vue";
export { default as NavbarEl } from "./NavbarEl.vue";
