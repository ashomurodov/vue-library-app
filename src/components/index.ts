export { default as Navbar } from "./NavBar.vue";
export { default as SearchBar } from "./SearchBar.vue";
export { default as bookItem } from "./bookItem.vue";
export { default as Loader } from "./LoadingItem.vue";
