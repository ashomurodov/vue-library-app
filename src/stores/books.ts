import { defineStore } from "pinia";
import { Mapper, iApi } from "@/service";
import type { SingleBook } from "@/types";
import { isTokenExpired } from "@/utils";
import router from "@/router";

interface State {
  books: SingleBook[];
  likedBooks: SingleBook[];
  similiarBook: SingleBook[];
  book: SingleBook;
  loading: boolean;
  search: string | null;
  hasFetchedFirstTime: boolean;
}

export const useBookStore = defineStore("bookStore", {
  state: (): State => ({
    books: [],
    similiarBook: [],
    likedBooks: [],
    book: {
      id: "",
      title: "",
      description: "",
      buyLink: "",
      publishDate: "",
      previewLink: "",
      image: "",
      authors: [],
      bookLink: "",
      isLiked: false,
    },
    loading: false,
    hasFetchedFirstTime: false,
    search: localStorage.getItem("lastSearch") ? localStorage.getItem("lastSearch") : "programming",
  }),
  actions: {
    async fetchBookListOfBooks() {
      if (!isTokenExpired()) {
        try {
          this.loading = true;
          const { data } = await iApi.fetchBooks(this.search!);
          console.log(data);
          this.books = data?.items?.map((item) => Mapper.Book(item));
        } catch (error) {
          console.log(error);
        } finally {
          this.loading = false;
        }
      } else if (isTokenExpired()) {
        alert("Token expired");
        router.push("/login");
      }
    },

    async fetchSingleBook(id: string) {
      if (!isTokenExpired() && this.books.length > 0) {
        const [singleBook] = this.books.filter((book) => book.id === id);
        this.book = singleBook;
      } else if (!isTokenExpired()) {
        try {
          this.loading = true;
          const { data } = await iApi.fetchSingleBook(id);
          this.book = Mapper.Book(data);
        } catch (error) {
          console.log(error);
        } finally {
          this.loading = false;
        }
      } else if (isTokenExpired()) {
        alert("Token is expired");
        router.push("/login");
      }
    },

    async fetchSimiliarBook(title: string) {
      if (!isTokenExpired()) {
        try {
          this.loading = true;
          // const {} = {}
        } catch (error) {
          console.log(error);
        } finally {
          this.loading = false;
        }
      }
    },

    addLikedBooks(id: string) {
      if (!isTokenExpired()) {
        const [likedBook] = this.books.filter((book) => book.id === id);
        if (likedBook.isLiked) {
          likedBook.isLiked = false;
          this.likedBooks = this.books.filter((book) => book.isLiked);
        } else {
          likedBook.isLiked = true;
          this.likedBooks = this.books.filter((book) => book.isLiked);
        }
      } else {
        alert("Your token expired pls resignIn");
        router.push("/login");
      }

      console.log(this.likedBooks);
    },
  },
});
