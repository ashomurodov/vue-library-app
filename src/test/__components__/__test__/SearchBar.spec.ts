import { render, screen } from "@testing-library/vue";
import { describe, it, expect, vi, test } from "vitest";
import { mount } from "@vue/test-utils";

import { SearchBar } from "../../../components";
import { useBookStore } from "@/stores/books";

describe("SearchBar", () => {
  const searchBarText = "Search";

  it("render text correctly", async () => {
    // arrange
    const searchId = "searchbarID";
    render(SearchBar, {
      props: { element: "button", id: searchId },
      slots: { default: searchBarText },
    });

    const search = await screen.findByText(searchBarText);
    expect(search.id).toBe(searchId);
    expect(search.textContent).toBe(searchBarText);
    expect(search.nodeName).toBe("BUTTON");
  });
});
