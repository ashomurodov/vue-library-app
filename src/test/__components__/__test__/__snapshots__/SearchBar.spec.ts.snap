// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > snap shot matches 1`] = `
{
  "baseElement": <body>
    <div>
      <div
        element="button"
        id="searchbarID"
      >
        <form
          class="flex gap-3 items-center"
        >
          <input
            class="outline-none border-2 rounded-lg w-28 sm:w-52 md:w-72 py-2 px-3 border-emerald-500"
            placeholder="Search.."
            type="text"
          />
          <button
            class="bg-emerald-500 rounded-lg w-16 sm:w-24 md:w-30 py-2 md:px-3 px-1 text-white"
            id="searchbarID"
          >
            Search
          </button>
        </form>
      </div>
    </div>
    <div>
      <div
        element="div"
      >
        <form
          class="flex gap-3 items-center"
        >
          <input
            class="outline-none border-2 rounded-lg w-28 sm:w-52 md:w-72 py-2 px-3 border-emerald-500"
            placeholder="Search.."
            type="text"
          />
          <button
            class="bg-emerald-500 rounded-lg w-16 sm:w-24 md:w-30 py-2 md:px-3 px-1 text-white"
            id="searchbarID"
          >
            Search
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      element="div"
    >
      <form
        class="flex gap-3 items-center"
      >
        <input
          class="outline-none border-2 rounded-lg w-28 sm:w-52 md:w-72 py-2 px-3 border-emerald-500"
          placeholder="Search.."
          type="text"
        />
        <button
          class="bg-emerald-500 rounded-lg w-16 sm:w-24 md:w-30 py-2 md:px-3 px-1 text-white"
          id="searchbarID"
        >
          Search
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
